### ISSUE TITLE
Add support for "Monitored users"

### ISSUE THREAD

**Issue by pajlada** [Chatterino/chatterino2#5208]
It should, as "Restricted users" has been requested in #5207 - have its color & "show in mentions" configurable

We currently omit monitored user messages since they come in over both IRC & PubSub, will need to be a bit nifty or somehow have it user-configurable to not make the messages show up messy

**Comment by squarebracket** [comment]
I'd like to take a crack at this. Since it comes in on both the IRC bridge and the eventsub, and the IRC bridge doesn't have that information annotated, we need to let the IRC message come in and then do some dedup based on the message ID.

Currently, all "low trust" (monitored/restricted) messages come in through eventsub, and two messages are created: a header (which shows whether it is a restricted/monitored message), and a body which is the message itself. Since the monitored messages come in through IRC too, the monitored messages are simply discarded, but the code is there to handle monitored messages through the eventsub code path.

I see a few different possibilities for the deduping:

1. Use `channel->replaceMessage` to swap out the IRC message for the header message, and then also `channel->addMessage(body)`. This could look a bit weird since you see the actual message get replaced with a header message (it's quick, but still noticeable).
2. Use `channel->replaceMessage` to swap out the IRC message for the properly styled body message, and insert the header message above it. This could also look weird, but at least the actual message body doesn't change too much.
3. Delete (or disable?) the IRC message and `channel->addMessage` the header + body.
4. Disregard a header, and simply `channel->replaceMessage` the IRC message with a properly styled message. This is probably the least visually janky solution, though it does mean the styling wouldn't be the same as restricted chat messages.

I'd love to get some feedback on what people think is the best solution.

**Comment by Nerixyz** [comment]
Here are some thoughts:

- I dislike 3, because then you'll have two messages with the same content - in many cases right next to each other.
- For channel point rewards, we already do some basic "upserting" (https://github.com/Chatterino/chatterino2/pull/5948) where we eagerly add a user's message and enrich it later once we get info about the reward. This is a bit different, since there, we already know from the IRC message, that it's a channel point redemption. Whereas here, we don't know whether a message is going to be from a monitored user.
- How should this look like in the usercard and search popups? Specifically, should the header be included in the message itself, or should they be two `Message`s? For reference, we do this for subscription messages where a subscription has a header and if the user typed a message, this is put into its own `Message`. For channel point rewards, on the other hand, we create one full message.
  I tend to prefer a single message (so 4), because we currently don't have a way of inserting a message at a specific position (this might be tricky because of filters).
- What if the message first comes through EventSub and is followed by IRC? Could keep a small queue to check against.
- When replacing the message, you'll probably have to rebuild it from scratch. We don't have functionality to clone messages (yet?).

**Comment by iProdigy** [comment]
another note: if streamer mode is active, ensure the message is still displayed but exclude any indication of the user being monitored

**Comment by squarebracket** [comment]
This has turned out to be a bit complex in terms of timing.

To get a little proof of concept going I combined [`makeSuspiciousUserMessageHeader`](https://github.com/Chatterino/chatterino2/blob/f080479c4f2afffd4f55edb84d5c7349aa12f4ef/src/providers/twitch/eventsub/MessageBuilder.cpp#L689) and [`makeSuspiciousUserMessageBody`](https://github.com/Chatterino/chatterino2/blob/f080479c4f2afffd4f55edb84d5c7349aa12f4ef/src/providers/twitch/eventsub/MessageBuilder.cpp#L757) to generate the single message needed for "replacement" (as an aside, the restricted user messages are currently split across two lines for searching, which this also fixed). This worked well enough:

https://github.com/user-attachments/assets/a0fc1796-d679-4719-9b27-bb1f916b309b

However, there is an issue with timing sometimes. Handling an eventsub message after an IRC message is completely processed is no problem, but it looks like the following event sequence can also happen:
1. IRC message received
2. eventsub message received, check for IRC message fails since it's not yet appended
3. `channel->addMessage` called with IRC message
4. `channel->addMessage` called with eventsub message

I'm not too sure what has its own thread here, it looks like the eventsub listener is in its own thread, is the irc listener in the same thread as the GUI or its own? In any case, it's a concurrency problem.

There's obviously a few different ways to handle this. Mutexing all message parsing seems like overkill, I feel like it'd make more sense to just handle deduping in a single thread closer to the actual append action. I did this in `Channel`'s `addMessage`:

```
    auto origMsg = this->messages_.rfind([message](MessagePtr m) {
        return m->id == message->id;
    });
    if (origMsg.has_value() && origMsg.value()->id != "") {
        if (message->flags.has(MessageFlag::MonitoredMessage)) {
            // we have a monitored message that should replace the non-flaired message
            this->replaceMessage(origMsg.value(), message);
            return; // don't trigger messageAppended.invoke
        } else {
            // we have a non-flaired message trying to replace a flaired message,
            // so do nothing
            return; // don't trigger messageAppended.invoke
        }
    }
```

which seems to work for all cases, but maybe that's a bit too tightly coupled for this base class...? `TwitchChannel::addMessage` seems a bit more logical, but I assume there's a reason `Channel::addMessage` is `final`.

I'm not too well-practised in C++ and you guys obviously know the codebase better, any suggestions for a good fix for this?

**Comment by Mm2PL** [comment]
> I'm not too sure what has its own thread here, it looks like the eventsub listener is in its own thread, is the irc listener in the same thread as the GUI or its own?

IRC I believe is on the gui thread.

There is a utility called `postToGuiThread` (`src/util/PostToThread.hpp`) which allows you to execute code on the next tick of Qt's event loop, it probably will help you with the concurrency issue.

We also do similar logic for replacing timeouts for richer variants. IRC delivers a `CLEARCHAT` message which is replaced by one received from eventsub which contains the moderator and reason. You probably could take inspiration from that.

**Comment by squarebracket** [comment]
Yeah, that's why I asked -- the sus user code is already using `postToGuiThread` for `channel->addMessage`. I suppose it was a bit silly of me not simply putting the logic in the lambda and seeing if that worked. Well, it did.

On to the next step.

**Comment by squarebracket** [comment]
I would like to do to something like this:

```
void Connection::onChannelSuspiciousUserMessage(
    const lib::messages::Metadata &metadata,
    const lib::payload::channel_suspicious_user_message::v1::Payload &payload)
{
    auto event = payload.event;
    // ...
    runInGuiThread([channel, time, msgId, event] {
        auto origMsg = channel->findMessageByID(msgId);
        if (origMsg)
        {
            auto replMsg = makeSuspiciousUserMessage(channel, time, event);
            channel->replaceMessage(origMsg, replMsg);
        }
        else
        {
            auto replMsg = makeSuspiciousUserMessage(channel, time, event);
            channel->addMessage(replMsg, MessageContext::Original);
        }
    });
}
```
I can create the replacement message outside the lambda, but I need to grab the thread data from the original message since it's not passed along with the suspicious user event. So I need both the event data and the original message data either inside or outside the lambda.

Is there an easy way to do this that I might be missing? My C++ knowledge isn't very deep.

**Comment by pajlada** [comment]
If you need data from the IRC message, I suspect you'll need to two disincts paths:
1. If the original message exists (i.e. if the suspicious message came in through IRC before it came in through eventsub): replace the message with combined information. This looks like the lambda you just posted.
2. If the original message does _not_ exist (i.e. if the suspicious message came in through eventsub before it came in through IRC): Put the eventsub message in some sort of pending map that the IRC message handler can check to build a full message when the message finally comes in through IRC.
